src/swupd/hash.go

```diff
@@ -1,17 +1,23 @@
 package swupd
 
+type hashval int
+
 // Hashes is a global map of indices to hashes
 var Hashes = []*string{}
+var invHash = make(map[string]hashval)
 
 // internHash adds only new hashes to the Hashes slice and returns the index at
 // which they are located
-func internHash(hash string) int {
-	for idx, val := range Hashes {
-		if *val == hash {
-			return idx
-		}
+func internHash(hash string) hashval {
+	if key, ok := invHash[hash]; ok {
+		return key
 	}
-
 	Hashes = append(Hashes, &hash)
-	return len(Hashes) - 1
+	key := hashval(len(Hashes) - 1)
+	invHash[hash] = key
+	return key
+}
+
+func (h hashval) String() string {
+	return *Hashes[int(h)]
```

Comment:
I think a range check here to avoid runtime panics would be a good add.

Reply:
What can we do if a range check fails?

Since the point of interning the values is that we will never call delete on them, if a range check fails then it shows either a bug in the code or someone has cast a value to a hashval. Either way the only sensible thing to do is abort the program, and at least a runtime panic gives you some useful information.

Reply:
I agree that we should fail, I just wasn't sure if a panic failure would be the best thing for this case. If you think panic is the right way to go that's fine but I usually prefer to avoid panics unless it is a 100% bug situation.

Reply:
I agree 100%. As I see it, we have to return a string. The only thing we could do differently is to say return fmt.Sprintf("Invalid hashval %d",int(h)), if the range check failed, but I think this will just make it too easy to ignore the problem. I think of the runtime panic being closer to an assert in C. if it fails we produce a corefile (which we then throw away so we can't debug the issue...)

Reply:
Okay great, thanks for the explanation!